tests/check_new_syntax.py

```diff
@@ -135,6 +148,27 @@ def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> None:
                 self.old_syntax_finder().visit(node.returns)
             self.generic_visit(node)
 
+        if not python_2_support_required:
+
+            def visit_ClassDef(self, node: ast.ClassDef) -> None:
+                context_manager_from_typing = self.module_imports_info_dict["context_manager_from_typing"]
+                async_context_manager_from_typing = self.module_imports_info_dict["async_context_manager_from_typing"]
+
+                for base in node.bases:
+                    match base:
+                        case ast.Subscript(value=ast.Name(id="ContextManager")) if context_manager_from_typing:
+                            raise_ContextManager_error(base)
+                        case ast.Subscript(value=ast.Name(id="AsyncContextManager")) if async_context_manager_from_typing:
+                            raise_AsyncContextManager_error(base)
+                        case ast.Name(id="ContextManager") if context_manager_from_typing:
+                            raise_ContextManager_error(base, is_subscript=False)
+                        case ast.Name(id="AsyncContextManager") if async_context_manager_from_typing:
+                            raise_AsyncContextManager_error(base, is_subscript=False)
+                        case _:
+                            pass
```

Comment:
Maybe it would make sense to run  `OldSyntaxFinder()` on the base classes? Maybe add an argument specifying whether it is for a base class or not, so that it doesn't complain about stuff that doesn't apply to base classes? Checking base classes too would also fix this comment `# TODO: can use built-in generics in type aliases`. If we do that, we probably don't need the `target_version` stuff anymore.

Reply:
That was my initial thought, as well, but I think it makes sense to keep the code separate. There are actually several things different about the code that needs to run on class definitions compared to the code that needs to run on annotations (though the code that needs to run on class definitions can be combined easily with the code that needs to run on `TypeVar` declarations -- my planned next PR 🙂.) So, I think I'd rather not use `OldSyntaxFinder()` here.

I am, however, happy to change this code block from using `match`/`case` syntax to using more backwards-compatible syntax, if you'd rather not bump `pyproject.toml` from py37 to py310. I personally find the `match`/`case` syntax much more readable and elegant in this case, but I can see why you might not want to bump the `target_version` in `pyproject.toml`.

Reply:
With regard to type aliases: yeah, I don't really want to touch them right now. The idea of figuring out which aliases are okay and which aren't looks kinda complicated and horrible.